build/gulpfile.reh.js

```diff
@@ -155,40 +181,49 @@ if (defaultNodeTask) {
 }
 
 function nodejs(platform, arch) {
-	const { remote } = require('./lib/gulpRemoteSource');
+	const { fetchUrls, fetchGithub } = require('./lib/fetch');
 	const untar = require('gulp-untar');
 
 	if (arch === 'ia32') {
 		arch = 'x86';
+	} else if (arch === 'armhf') {
+		arch = 'armv7l';
+	} else if (arch === 'alpine') {
+		platform = 'alpine';
+		arch = 'x64';
 	}
 
-	if (platform === 'win32') {
-		if (product.nodejsRepository) {
-			log(`Downloading node.js ${nodeVersion} ${platform} ${arch} from ${product.nodejsRepository}...`);
-			return assetFromGithub(product.nodejsRepository, nodeVersion, name => name === `win-${arch}-node.exe`)
-				.pipe(rename('node.exe'));
-		}
-		log(`Downloading node.js ${nodeVersion} ${platform} ${arch} from https://nodejs.org`);
-		return remote(`/dist/v${nodeVersion}/win-${arch}/node.exe`, { base: 'https://nodejs.org', verbose: true })
-			.pipe(rename('node.exe'));
-	}
+	log(`Downloading node.js ${nodeVersion} ${platform} ${arch} from ${product.nodejs.repository}...`);
+
+	const checksumSha256 = getNodeChecksum(nodeVersion, platform, arch);
 
-	if (arch === 'alpine' || platform === 'alpine') {
-		const imageName = arch === 'arm64' ? 'arm64v8/node' : 'node';
-		log(`Downloading node.js ${nodeVersion} ${platform} ${arch} from docker image ${imageName}`);
-		const contents = cp.execSync(`docker run --rm ${imageName}:${nodeVersion}-alpine /bin/sh -c 'cat \`which node\`'`, { maxBuffer: 100 * 1024 * 1024, encoding: 'buffer' });
-		return es.readArray([new File({ path: 'node', contents, stat: { mode: parseInt('755', 8) } })]);
+	if (checksumSha256) {
+		log(`Using SHA256 checksum for checking integrity: ${checksumSha256}`);
+	} else {
+		log.warn(`Unable to verify integrity of downloaded node.js binary because no SHA256 checksum was found!`);
 	}
 
-	if (arch === 'armhf') {
-		arch = 'armv7l';
+	switch (platform) {
+		case 'win32':
+			if (product.nodejs.repository !== 'https://nodejs.org') {
+				return fetchGithub(product.nodejs.repository, { version: product.nodejs.version, name: `win-${arch}-node.exe`, checksumSha256 }).pipe(rename('node.exe'));
+			} else {
+				return fetchUrls(`/dist/v${nodeVersion}/win-${arch}/node.exe`, { base: 'https://nodejs.org', checksumSha256 }).pipe(rename('node.exe'));
+			}
+		case 'darwin':
+		case 'linux':
+			return fetchUrls(`/dist/v${nodeVersion}/node-v${nodeVersion}-${platform}-${arch}.tar.gz`, { base: 'https://nodejs.org', checksumSha256 })
+				.pipe(flatmap(stream => stream.pipe(gunzip()).pipe(untar())))
+				.pipe(filter('**/node'))
+				.pipe(util.setExecutableBit('**'))
+				.pipe(rename('node'));
+		case 'alpine':
+			if (product.nodejs.repository !== 'https://nodejs.org') {
+				return fetchGithub(product.nodejs.repository, { version: product.nodejs.version, name: `alpine-${arch}-node`, checksumSha256 }).pipe(rename('node'));
+			} else {
+				throw new Error('There is no official release of node.js for Alpine Linux');
```

Comment:
Why are we breaking this case ? We have been providing the node binaries from the [node docker repo](https://hub.docker.com/r/arm64v8/node/) which is actually published by https://github.com/nodejs/docker-node. Maybe for sha in this case, we can rely on the sha of the docker image ? which is available at https://hub.docker.com/layers/arm64v8/node/16.17.1-alpine/images/sha256-0b4e643f8ae73b784e5366e558e11cfbabb727020032e6665a0cd85a3ec48b3a?context=explore

Reply:
The intent is to avoid breaking the build when the docker image changes. @joaomoreno was not objecting when I asked him about it.

Reply:
This will also break the production build, since there is no alpine builds available from `vscode-node`.

> The intent is to avoid breaking the build when the docker image changes

We can lock the docker sha in that case and compare with ` docker inspect --format='{{index .RepoDigests 0}}' <imagename>`

Reply:
No, I uploaded alpine node images to our repo, so that should be fine. But I like your idea as well.

@joaomoreno thoughts?

Reply:
Given we are not yet building the alpine versions in `vscode-node`, it would be good to get the builds from the actual source (dockerhub in this case) and have sha checks for that. I am good with serving from `vscode-node` if we end up building the image.

Reply:
Went back to using docker image but still do checksum verification.